### DISCUSSION TITLE
Test fails, reverts without data

### DISCUSSION THREAD

**Discussion by omeguhh** [foundry-rs/foundry#2449] [Help] [answered]
Hello, using the Stargate contracts for the first time and have been running into issues with it during tests. 

**test**
``` solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "../src/Contract.sol";
import "forge-std/Test.sol";

contract StargateFarmingTest is Test {
    StargateFarming stargate;   
    IEthRouter ethRouter = IEthRouter(0xbf22f0f184bCcbeA268dF387a49fF5238dD23E40);
    address public addrTest = 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266;


    function setUp() public {
        stargate = new StargateFarming(ethRouter);
    }

    function testCanAddLiquidity() public {
        vm.startPrank(addrTest);
        stargate.addLiquidity{value:1 ether}();
    }
}
```

**contract**
```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.0;

import "./Interfaces/IEthRouter.sol";



contract StargateFarming{


    // uint16 public immutable poolId;
    // address public immutable ethVault;
    // address public immutable stgRouter;
    
    IEthRouter ethRouter;

    constructor(IEthRouter _ethRouter){
        ethRouter = _ethRouter;
    }



    function addLiquidity() public payable {
        ethRouter.addLiquidityETH();
    }

    receive() external payable {}


}
```

It's not much because Im just trying to test with it. Another dev I work with mentioned it worked with Hardhat. When i run forge test I get fail reason revert. When i do cast send with the contract address and use the same from address, the value is sending, so Im unsure of what I am doing wrong or if this is a forge issue. Appreciate the assistance!

Here is the stargate contract I made an interface from: https://arbiscan.io/address/0xbf22f0f184bccbea268df387a49ff5238dd23e40#code

**Comment by onbjerg** [comment, marked as answer]
What command are you running?

Edit: Solution is

> You need to specify a fork URL (and a block number) to be able to interact with live contracts (`--fork-url <rpc url>` and `--fork-block-number <number>` respectively)

**Reply by omeguhh** [reply]
Hey @onbjerg long time no see! I ran three.
1.  forge test -vvvv to see if i could diagnose the problem
2. cast call 0xbf22f0f184bCcbeA268dF387a49fF5238dD23E40 "addLiquidityETH()" 
("addLiquidityETH()" is one of the only functions on the contract. It is the only function I use from the interface I created.)
3. cast send 0xbf22f0f184bCcbeA268dF387a49fF5238dD23E40 --from 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266 --value 1ether (From is an address provided by anvil)


Please let me know if I need to clarify on anything else. Thank you ser!

**Reply by onbjerg** [reply]
You need to specify a fork URL (and a block number) to be able to interact with live contracts (`--fork-url <rpc url>` and `--fork-block-number <number>` respectively)

**Reply by omeguhh** [reply]
Hey thanks for the answer. Would the `--fork-url` be something like an alchemy key or would it be for example, arbitrum rinkebys rpc url? I also noticed in the Foundry book the example showed `--fork-block-number 1` . Would 1 work or should I use the block that shows up when I check when the contract was created? Thanks so much
![Screen Shot 2022-07-24 at 2 18 15 PM](https://user-images.githubusercontent.com/105019014/180662531-55cf0d93-909e-46ea-bdbd-10f897d9dc41.png)

**Reply by onbjerg** [reply]
The fork URL is a fully formed RPC URL, so for Alchemy it would look something like `https://<network>.alchemyapi.io/v2/<key>`. If you don't know what block number to use, I would advise using the latest. Using one too far in the past can be slow, and using the block number the contract was created on likely won't give you any data (since no one has used the contract yet). The reason for setting the block number and not letting Forge automatically use the latest on every test run is that specifying the block number lets Forge cache the response data from the node and speed up the test after the first run.

**Reply by omeguhh** [reply]
Yeah my apologies, that was what I was referring to in regards to the alchemy question. Sorry for the confusion. Just wanted to make sure that was what I was supposed to use. 

Thanks so much! Just tried that and my problem is solved. However, I had to also add `ethRouter.addLiquidityETH{value: msg.value}();` in my contracts function because without the `{value: msg.value}` bit the Stargate contract was reverting with Stargate: msg.value is 0. Not 100% sure why that is, but it works now. Thanks again @onbjerg you're a king

**Reply by onbjerg** [reply]
That's because the eth value sent to your `addLiquidity` function is not automatically sent to any function you call - you must be explicit about it in your contract :) Glad to be of help

**Reply by omeguhh** [reply]
![Screen Shot 2022-07-24 at 2 40 51 PM](https://user-images.githubusercontent.com/105019014/180663190-b0168049-66e5-45e1-86a0-98f86a62e430.png)

Yeah that makes sense. First time that i've worked with a contract like this and since it doesnt take any arguments it tripped me up a bit. But yeah anyways, really appreciate it and hope you enjoy your weekend. I know I will now that I'm over this hump
